### ISSUE TITLE
jattach: No such file or directory from a container

### ISSUE THREAD

**Issue by fcamblor** [jvm-profiling-tools/async-profiler#207]
Hi,

I'm just trying to install async-profiler on a docker instance, but for some strange reasons, it doesn't seem to work when I try to run it at will.

Dockerfile I'm using to install async-profiler & flamegraph :
```
FROM tomcat:7-jre8-alpine

RUN rm -rf /usr/local/tomcat/webapps/ROOT

COPY current-app.war /usr/local/tomcat/webapps/ROOT.war
COPY setenv.sh /usr/local/tomcat/bin/setenv.sh

# Installing async-profiler 1.5
RUN mkdir /usr/local/async-profiler/ &&\
      wget -O /usr/local/async-profiler/async-profiler.tar.gz https://github.com/jvm-profiling-tools/async-profiler/releases/download/v1.5/async-profiler-1.5-linux-x64.tar.gz &&\
      cd /usr/local/async-profiler/ &&\
      tar -xvzf async-profiler.tar.gz &&\
      rm -f /usr/local/async-profiler/async-profiler.tar.gz

# Installing flamegraphs
RUN mkdir /usr/local/flame-graph/ &&\
    wget -O /usr/local/flame-graph/flame-graph.zip https://github.com/brendangregg/FlameGraph/archive/1b1c6deede9c33c5134c920bdb7a44cc5528e9a7.zip &&\
    cd /usr/local/flame-graph/ && unzip flame-graph.zip &&\
    cd /usr/local/flame-graph/ &&\
    mv FlameGraph-1b1c6deede9c33c5134c920bdb7a44cc5528e9a7/* ./ &&\
    rm -Rf FlameGraph-1b1c6deede9c33c5134c920bdb7a44cc5528e9a7 &&\
    rm -f /usr/local/flame-graph/flame-graph.zip

# Running tomcat as this is the standard container execution
CMD ["sh", "-c", "cd /usr/local/tomcat/ && catalina.sh run"]
```

Once built (with `security_opt=seccomp:unconfined`), when I `docker exec` into the container, and try to run `/usr/loca/async-profiler/profiler.sh`, I get `/usr/local/async-profiler/build/jattach: No such file or directory` :
```
docker exec -ti f8168fed6fce /usr/local/async-profiler/profiler.sh -d 20 -o collapsed -f /tmp/collapsed.txt 1 && /usr/local/flame-graph/flamegraph.pl --colors=java /tmp/collapsed.txt
/usr/local/async-profiler/profiler.sh: line 61: /usr/local/async-profiler/build/jattach: No such file or directory
```

Although a `more /usr/local/async-profiler/build/jattach` works perfectly.

I even tried to change permissions on jattach and surrounding directories, but it doesn't help.
Do you have any idea on why `jattach` doesn't seem to work properly ?

**Comment by fcamblor** [comment]
Found my problem, seems like the `No such file or directory` error is not an accurate error message when at least one required dependency is missing on a docker instance (occuring frequently on alpine instances)

More infos on [this (useful!) post](https://superuser.com/a/1176328)

I solved my issue by running an `RUN apk add libc6-compat` in my dockerfile

**Comment by apangin** [comment]
@fcamblor Thank you for the information - this is not useless noise at all :)
There is already a separate package with jattach for Alpine Linux. I'll also consider creating a package with async-profiler (which does not require extra dependencies).

I noticed your [profiling script](https://gist.github.com/fcamblor/1dc5a4edc27f89636b473e30c4efa7d2#file-generate-async-profiler-flamechart-sh) uses `flamegraph.pl`. Actually it's not needed, since async-profiler can generate flame graphs on its own - just specify output file with `.svg` extension. This can be helpful for reducing container image size, as `perl` package will not be needed either.

It's strange that perf_events don't work. What error message does profiler print in this case? Try disabling [AppArmor](https://docs.docker.com/engine/security/apparmor/) and make sure `perf_event_paranoid` is set to `1`.

**Comment by fcamblor** [comment]
> I noticed your profiling script uses flamegraph.pl. Actually it's not needed, since async-profiler can generate flame graphs on its own - just specify output file with .svg extension.

Wow cool, never known of the special `.svg` extension, that's nice ! :-)

About `perf_events`, take it with a grain of salt as it may be due to misconfiguration on my side.
I tried several `docker-compose` options to put `security_opt=seccomp:unconfined` but there are good chances I failed at it.
I'm not good at this AppArmor stuff, and when I tried to set `perf_event_paranoid` to `1` at some point, it didn't worked (but I can't be 100% sure it didn't worked due to something else, I would have to try it again)

Anyway, falling back to `-e itimer` as stated on the doc made it work so, at least for the moment, I'm OK with that even if not 100% in the spirit of `async-profiler`.

BTW, thanks for the great work you did on `async-profiler`, it saved me a couple of hours, multiple times ❤️
